block/manager.go

```diff
@@ -229,6 +229,22 @@ func (m *Manager) Start(ctx context.Context) error {
 	}
 	amIProposer := amIProposerOnSL || m.AmIProposerOnRollapp()
 
+	if m.forkNeeded() {
```

Comment:
if there is nothing to do with fullnode, i would move this part the runAsProposer func, probably after shouldRotate, to make sure the node is a proposer, and is fully synced before starting producing blocks. i would also avoid loading the instruction twice.